File: seed/problems/005_binary-tree-level-order/solutions/solution.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine().trim());
        StringTokenizer st = new StringTokenizer(br.readLine());
        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
            a[i] = Integer.parseInt(st.nextToken());
        }

        // BFS approach: process level by level using index-based traversal
        // Node at index i has left child at 2i+1 and right child at 2i+2
        Deque<Integer> queue = new ArrayDeque<>();
        if (n > 0 && a[0] != -1) {
            queue.addLast(0);
        }

        StringBuilder sb = new StringBuilder();
        boolean firstLine = true;
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            if (!firstLine) sb.append('\n');
            firstLine = false;

            for (int i = 0; i < levelSize; i++) {
                int idx = queue.pollFirst();
                if (i > 0) sb.append(' ');
                sb.append(a[idx]);

                int left = 2 * idx + 1;
                int right = 2 * idx + 2;
                if (left < n && a[left] != -1) {
                    queue.addLast(left);
                }
                if (right < n && a[right] != -1) {
                    queue.addLast(right);
                }
            }
        }

        System.out.println(sb);
    }
}

File: seed/problems/005_binary-tree-level-order/solutions/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<int> a(n);
    for (int i = 0; i < n; i++) {
        cin >> a[i];
    }

    // BFS approach: process level by level using index-based traversal
    // Node at index i has left child at 2i+1 and right child at 2i+2
    queue<int> q;
    if (n > 0 && a[0] != -1) {
        q.push(0);
    }

    bool first_line = true;
    while (!q.empty()) {
        int level_size = q.size();
        vector<int> level_values;

        for (int i = 0; i < level_size; i++) {
            int idx = q.front();
            q.pop();
            level_values.push_back(a[idx]);

            int left = 2 * idx + 1;
            int right = 2 * idx + 2;
            if (left < n && a[left] != -1) {
                q.push(left);
            }
            if (right < n && a[right] != -1) {
                q.push(right);
            }
        }

        if (!first_line) cout << '\n';
        for (int i = 0; i < (int)level_values.size(); i++) {
            if (i > 0) cout << ' ';
            cout << level_values[i];
        }
        first_line = false;
    }

    cout << '\n';
    return 0;
}
